urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/response/successful/OK.java

```diff
@@ -0,0 +1,698 @@
+	    // This might throw one of these exceptions:
+	    //  - ResourceNotFoundException
+	    //  - AuthorizationException
+	    //  - HeaderFormatException
+	    //  - DataFormatException
+	    //  - MissingResourceException
+	    //  - ConfigurationException
+	    //  - SecurityException
+	    //  - IOException
+	    Resource resource = this.getHTTPHandler().getResourceAccessor().locate( uri,
+										    this.getRequestHeaders(),
+										    this.getPostData(),
+										    additionalResourceSettings,
+										    optionalReturnSettings,
+										    this.getSessionID()
+										    );
+	    this.getHTTPHandler().getLogger().log( Level.INFO,
+						   getClass().getName() + ".prepare(...)",
+						   "Resource accessor granted access to the file '" + uri.getPath() + "'. optionalReturnSettings=" + optionalReturnSettings );
+
+
+	    resource.getReadLock().lock();
+	    resource.open( true ); // Open in read-only mode 
+
+
+	    // Supported since version 1.0.2.alpha:
+	    // The 'Content-Range' header
+	    String request_contentRange = this.getRequestHeaders().getStringValue( HTTPHeaders.NAME_CONTENT_RANGE );
+	    if( request_contentRange != null ) {
+		
+		// This might throw a MalformedRequestException
+		// (NullPointerException cannot be raised because the header field is not null)
+		ContentRange cRange = ContentRange.parse( request_contentRange );
+		// The numerical fields are in range so far ...
+		// ... but we have to check if the instanceLength is in range (the parser does not
+		//     know the resource's length).
+		if( (cRange.getInstanceLength() != -1 
+		     && cRange.getInstanceLength() != resource.getLength()
+		     )
+		    || cRange.getLastBytePosition() >= resource.getLength() ) {
+		    
+		    // TODO:
+		    //  Server should send a '416 Requested range not satisfiable' error response!
+		    optionalReturnSettings.put( Constants.KEY_GENERATED_STATUS_CODE, 
+						new BasicNumberType(Constants.HTTP_STATUS_CLIENTERROR_REQUEST_RANGE_NOT_SATISFIABLE) );
+		    optionalReturnSettings.put( Constants.KEY_GENERATED_REASON_PHRASE,
+						new BasicStringType("Requested range not satisfiable") );
+		    //throw new MalformedRequestException( "Cannot satisfy requested range (out of bounds)" );
+		    throw new ParametrizedHTTPException( "Cannot satisfy requested range (out of bounds).",
+							 Constants.HTTP_STATUS_CLIENTERROR_REQUEST_RANGE_NOT_SATISFIABLE,
+							 "Requested range not satisfiable" );
+							 
+
+		}
+		
+		// Build a 'ranged' resource.
+		// ... send a '206 Partial Content' response.
+		optionalReturnSettings.put( Constants.KEY_GENERATED_STATUS_CODE, 
+					    new BasicNumberType(Constants.HTTP_STATUS_SUCCESSFUL_PARTIAL_CONTENT) );
+		optionalReturnSettings.put( Constants.KEY_GENERATED_REASON_PHRASE,
+					    new BasicStringType("Partial Content") );
+		long oldResourceLength = resource.getLength();
+		resource = new RangedResource( resource,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *resource*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).